realm/realm-library/src/main/cpp/objectserver_shared.hpp

```diff
@@ -81,6 +82,9 @@ class JniSession {
                             local_env, obj, "notifySessionError", "(ILjava/lang/String;)V");
                     local_env->CallVoidMethod(
                             obj, notify_error_handler, error_code.value(), local_env->NewStringUTF(message.c_str()));
+                    if (local_env->ExceptionCheck() == JNI_TRUE) {
+                        throw realm::jni_util::JniPendingException("Exception in error handler");
```

Comment:
It is very hard to catch an exception from a callback (different stack, etc.) so I think that the app will terminate in most cases.  See for example http://stackoverflow.com/questions/13506900/java-asynchronous-exceptions-can-i-catch-them